File: git/go/diff.go
package mygit

// diff (SPEC.md §11) — 두 줄 목록 사이의 가장 짧은 편집 스크립트.
//
// 세 단계다. (1) 앞뒤의 같은 줄을 떼어 둔다. (2) 남은 가운데서 Myers
// 의 탐욕 탐색으로 가장 짧은 스크립트를 찾는다. (3) 바뀐 줄 묶음을
// git 처럼 위아래로 밀어 자리를 정한다 — 같은 줄이 되풀이되는 곳에서는
// 어디를 바뀐 줄로 칠지가 여럿이라, 이 단계가 없으면 git 과 덩어리
// 자리가 달라진다.
//
// 줄은 string(바이트 열)이고 줄바꿈까지 품는다. rchg 는 줄마다
// "바뀌었나" 의 bool 목록이다.

import (
	"bytes"
	"fmt"
	"os"
	"strings"
)

const (
	context     = 3
	binaryProbe = 8000
)

// SplitLines 는 바이트 → 줄 목록. 줄마다 '\n' 을 품고, 마지막 줄만
// 없을 수 있다.
func SplitLines(data []byte) []string {
	var lines []string
	for len(data) > 0 {
		i := bytes.IndexByte(data, '\n')
		if i < 0 {
			lines = append(lines, string(data))
			break
		}
		lines = append(lines, string(data[:i+1]))
		data = data[i+1:]
	}
	return lines
}

// ── 2 단계: Myers 앞방향 탐욕 탐색 (SPEC.md §11.2) ─────────────────

// forward 는 가운데 a·b 의 (ra, rb). 대각선 k 마다 가장 멀리 간 x 를
// v[k] 에 둔다. d 마다 v 의 [-d, d] 조각을 남겨 두었다가 (N, M) 에서
// 거꾸로 같은 판정을 되밟아 편집을 표시한다.
// O((N+M)·D) 시간, O(D²) 공간.
func forward(a, b []string) ([]bool, []bool) {
	n, m := len(a), len(b)
	off := n + m + 1
	v := make([]int, 2*off+1)
	var trace [][]int
	for d := 0; d <= n+m; d++ {
		trace = append(trace, append([]int(nil), v[off-d:off+d+1]...))
		for k := -d; k <= d; k += 2 {
			var x int
			if k == -d || (k != d && v[off+k-1] < v[off+k+1]) {
				x = v[off+k+1] // 아래로: b 의 줄을 끼움
			} else {
				x = v[off+k-1] + 1 // 오른쪽: a 의 줄을 지움
			}
			y := x - k
			for x < n && y < m && a[x] == b[y] {
				x, y = x+1, y+1
			}
			v[off+k] = x
			if x >= n && y >= m {
				return backtrack(trace, n, m, d)
			}
		}
	}
	panic("Myers 탐색이 끝나지 않았다")
}

func backtrack(trace [][]int, n, m, dfin int) ([]bool, []bool) {
	ra, rb := make([]bool, n), make([]bool, m)
	x, y := n, m
	for d := dfin; d > 0; d-- {
		v := func(k int) int { return trace[d][k+d] }
		k := x - y
		down := k == -d || (k != d && v(k-1) < v(k+1))
		pk := k - 1
		if down {
			pk = k + 1
		}
		px := v(pk)
		py := px - pk
		if down {
			rb[py] = true
		} else {
			ra[px] = true
		}
		x, y = px, py
	}
	return ra, rb
}

// Myers 는 1·2 단계 — 앞뒤를 깎고 가운데를 앞방향 Myers 로.
func Myers(a, b []string) ([]bool, []bool) {
	n, m := len(a), len(b)
	s := 0
	for s < n && s < m && a[s] == b[s] {
		s++
	}
	e := 0
	for e < n-s && e < m-s && a[n-1-e] == b[m-1-e] {
		e++
	}
	ma, mb := forward(a[s:n-e], b[s:m-e])
	ra, rb := make([]bool, n), make([]bool, m)
	copy(ra[s:], ma)
	copy(rb[s:], mb)
	return ra, rb
}

// ── 3 단계: 밀어 붙이기 (git 의 xdl_change_compact, 휴리스틱 없이) ──

// group 은 바뀐 줄 묶음 [start, end). 빈 묶음(start == end)도 자리다.
type group struct {
	chg        []bool
	n          int // 끝의 가짜 줄 하나를 뺀 길이
	start, end int
}

func newGroup(chg []bool) *group {
	g := &group{chg: chg, n: len(chg) - 1}
	for g.end < g.n && chg[g.end] {
		g.end++
	}
	return g
}

func (g *group) next() bool {
	if g.end == g.n {
		return false
	}
	g.end++
	g.start = g.end
	for g.end < g.n && g.chg[g.end] {
		g.end++
	}
	return true
}

func (g *group) previous() bool {
	if g.start == 0 {
		return false
	}
	g.start--
	g.end = g.start
	for g.start > 0 && g.chg[g.start-1] {
		g.start--
	}
	return true
}

func (g *group) slideDown(recs []string) bool {
	if g.end < g.n && recs[g.start] == recs[g.end] {
		g.chg[g.start], g.chg[g.end] = false, true
		g.start, g.end = g.start+1, g.end+1
		for g.end < g.n && g.chg[g.end] {
			g.end++
		}
		return true
	}
	return false
}

func (g *group) slideUp(recs []string) bool {
	if g.start > 0 && recs[g.start-1] == recs[g.end-1] {
		g.start, g.end = g.start-1, g.end-1
		g.chg[g.start], g.chg[g.end] = true, false
		for g.start > 0 && g.chg[g.start-1] {
			g.start--
		}
		return true
	}
	return false
}

// compact 는 한 쪽 파일의 바뀐 줄 묶음을 밀어 자리를 정한다(§11.2).
// 묶음마다 위로 끝까지, 다시 아래로 끝까지 민다(밀다가 이웃 묶음과
// 붙으면 처음부터). 상대 파일의 바뀐 묶음과 끝이 맞는 자리가
// 있었으면 그리로 되올리고, 없으면 맨 아래에 둔다. 상대 쪽 묶음 표지
// go 는 묶음과 발을 맞춰 움직인다. O(줄 수 × 미는 거리).
func compact(recs []string, rchg []bool, ochg []bool) []bool {
	chg := append(append([]bool(nil), rchg...), false)
	g := newGroup(chg)
	gother := newGroup(append(append([]bool(nil), ochg...), false))
	for {
		if g.end != g.start {
			var earliest, matchEnd int
			for {
				size := g.end - g.start
				matchEnd = -1
				for g.slideUp(recs) {
					gother.previous()
				}
				earliest = g.end
				if gother.end > gother.start {
					matchEnd = g.end
				}
				for g.slideDown(recs) {
					gother.next()
					if gother.end > gother.start {
						matchEnd = g.end
					}
				}
				if size == g.end-g.start {
					break
				}
			}
			if g.end != earliest && matchEnd != -1 {
				for gother.end == gother.start {
					g.slideUp(recs)
					gother.previous()
				}
			}
		}
		if !g.next() {
			break
		}
		gother.next()
	}
	return chg[:len(chg)-1]
}

// EditFlags 는 세 단계를 다 거친 (ra, rb) — 계약의 전부(§11.2).
func EditFlags(a, b []string) ([]bool, []bool) {
	ra, rb := Myers(a, b)
	ra = compact(a, ra, rb)
	rb = compact(b, rb, ra)
	return ra, rb
}

// Change 는 바뀐 곳 하나 — a·b 의 자리와 줄 수.
type Change struct{ A, B, NA, NB int }

// BuildChanges 는 바뀐 곳들, 앞에서부터. 끝에서 앞으로 훑으며 같은
// 자리에서 만나는 지운 묶음과 끼운 묶음을 한 바뀐 곳으로 묶는다
// (git 의 xdl_build_script).
func BuildChanges(ra, rb []bool) []Change {
	var out []Change
	i1, i2 := len(ra), len(rb)
	for i1 > 0 || i2 > 0 {
		if (i1 > 0 && ra[i1-1]) || (i2 > 0 && rb[i2-1]) {
			l1, l2 := i1, i2
			for i1 > 0 && ra[i1-1] {
				i1--
			}
			for i2 > 0 && rb[i2-1] {
				i2--
			}
			out = append(out, Change{i1, i2, l1 - i1, l2 - i2})
		} else {
			i1, i2 = i1-1, i2-1
		}
	}
	for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
		out[i], out[j] = out[j], out[i]
	}
	return out
}

// isFunc 는 git 기본 드라이버의 함수 줄 — 첫 바이트가 영문자·'_'·'$'.
func isFunc(line string) bool {
	if line == "" {
		return false
	}
	c := line[0]
	return c|0x20 >= 'a' && c|0x20 <= 'z' || c == '_' || c == '$'
}

func span(start, count int) string {
	first := start
	if count > 0 {
		first++
	}
	if count == 1 {
		return fmt.Sprint(first)
	}
	return fmt.Sprintf("%d,%d", first, count)
}

const asciiSpace = " \t\n\r\v\f"

// UnifiedDiff 는 덩어리들(SPEC.md §11.3). 같으면 빈 것.
func UnifiedDiff(a, b []string) []byte {
	ra, rb := EditFlags(a, b)
	ch := BuildChanges(ra, rb)
	var out []string
	for i := 0; i < len(ch); {
		j := i
		for j+1 < len(ch) && ch[j+1].A-(ch[j].A+ch[j].NA) <= 2*context {
			j++
		}
		first, last := ch[i], ch[j]
		s1, s2 := max(first.A-context, 0), max(first.B-context, 0)
		e1 := min(last.A+last.NA+context, len(a))
		e2 := min(last.B+last.NB+context, len(b))
		fn := ""
		for q := s1 - 1; q >= 0; q-- {
			if isFunc(a[q]) {
				l := strings.TrimRight(a[q], asciiSpace)
				fn = " " + strings.TrimRight(l[:min(len(l), 80)],
					asciiSpace)
				break
			}
		}
		out = append(out, fmt.Sprintf("@@ -%s +%s @@%s\n",
			span(s1, e1-s1), span(s2, e2-s2), fn))
		p1 := s1
		for _, c := range ch[i : j+1] {
			for q := p1; q < c.A; q++ {
				out = append(out, " "+a[q])
			}
			for q := c.A; q < c.A+c.NA; q++ {
				out = append(out, "-"+a[q])
			}
			for q := c.B; q < c.B+c.NB; q++ {
				out = append(out, "+"+b[q])
			}
			p1 = c.A + c.NA
		}
		for q := p1; q < e1; q++ {
			out = append(out, " "+a[q])
		}
		i = j + 1
	}
	var buf bytes.Buffer
	for _, line := range out {
		buf.WriteString(line)
		if !strings.HasSuffix(line, "\n") {
			buf.WriteString("\n\\ No newline at end of file\n")
		}
	}
	return buf.Bytes()
}

// Side 는 diff 의 한 쪽 — (모드, 이름, 바이트).
type Side struct {
	Mode uint32
	Oid  string
	Data []byte
}

// FileDiff 는 파일 하나의 diff 전체(SPEC.md §11.4). old·new 가 nil
// 이면 새로 생김·지워짐. 같으면 빈 것.
func FileDiff(pa, pb string, old, new *Side) []byte {
	if old != nil && new != nil && old.Mode == new.Mode &&
		old.Oid == new.Oid {
		return nil
	}
	qa, qb := QuotePath("a/"+pa, false), QuotePath("b/"+pb, false)
	rows := []string{"diff --git " + qa + " " + qb}
	z := "0000000"
	switch {
	case old == nil:
		rows = append(rows, fmt.Sprintf("new file mode %06o",
			new.Mode), "index "+z+".."+new.Oid[:7])
	case new == nil:
		rows = append(rows, fmt.Sprintf("deleted file mode %06o",
			old.Mode), "index "+old.Oid[:7]+".."+z)
	default:
		if old.Mode != new.Mode {
			rows = append(rows, fmt.Sprintf("old mode %06o", old.Mode),
				fmt.Sprintf("new mode %06o", new.Mode))
		}
		if old.Oid == new.Oid {
			return []byte(strings.Join(rows, "\n") + "\n") // 모드만
		}
		idx := "index " + old.Oid[:7] + ".." + new.Oid[:7]
		if old.Mode == new.Mode {
			idx += fmt.Sprintf(" %06o", old.Mode)
		}
		rows = append(rows, idx)
	}
	var da, db []byte
	na, nb := "/dev/null", "/dev/null"
	if old != nil {
		da, na = old.Data, qa
	}
	if new != nil {
		db, nb = new.Data, qb
	}
	if bytes.IndexByte(da[:min(len(da), binaryProbe)], 0) >= 0 ||
		bytes.IndexByte(db[:min(len(db), binaryProbe)], 0) >= 0 {
		rows = append(rows, "Binary files "+na+" and "+nb+" differ")
		return []byte(strings.Join(rows, "\n") + "\n")
	}
	rows = append(rows, "--- "+na, "+++ "+nb)
	return append([]byte(strings.Join(rows, "\n")+"\n"),
		UnifiedDiff(SplitLines(da), SplitLines(db))...)
}

// BlobSide 는 저장소의 blob 에서 한 쪽을 만든다.
func BlobSide(gitdir string, b Blob) (*Side, error) {
	_, data, err := ReadObject(gitdir, b.Oid)
	if err != nil {
		return nil, err
	}
	return &Side{b.Mode, b.Oid, data}, nil
}

// DiskSide 는 디스크의 파일에서 한 쪽을 만든다. 없으면 nil.
func DiskSide(path string) *Side {
	st, err := os.Stat(path)
	if err != nil || !st.Mode().IsRegular() {
		return nil
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	mode := uint32(0o100644)
	if st.Mode()&0o100 != 0 {
		mode = 0o100755
	}
	return &Side{mode, HashObject("blob", data), data}
}

File: git/ts/src/diff.ts
// diff (SPEC.md §11) — 두 줄 목록 사이의 가장 짧은 편집 스크립트.
//
// 세 단계다. (1) 앞뒤의 같은 줄을 떼어 둔다. (2) 남은 가운데서 Myers 의
// 탐욕 탐색으로 가장 짧은 스크립트를 찾는다. (3) 바뀐 줄 묶음을 git
// 처럼 위아래로 밀어 자리를 정한다 — 같은 줄이 되풀이되는 곳에서는
// 어디를 바뀐 줄로 칠지가 여럿이라, 이 단계가 없으면 git 과 덩어리
// 자리가 달라진다. 다섯 언어가 같은 세 단계를 밟는다.
//
// 줄은 바이트 문자열(latin1)이고 줄바꿈까지 품는다 — 끝 줄바꿈이 없는
// 줄은 있는 줄과 다른 줄이다. === 가 곧 바이트 비교다. rchg 는 줄마다
// "바뀌었나" 의 boolean 배열이다. 출력도 바이트 문자열이다.
import * as objects from './objects';
import { quotePath } from './worktree';

const CONTEXT = 3;
const BINARY_PROBE = 8000;

// 파일 하나의 한쪽 — [모드, blob 이름, 바이트]. 없으면 null.
export type Side = [mode: number, oid: string, data: Buffer];
// 바뀐 곳 — [a 자리, b 자리, a 줄 수, b 줄 수]
export type Change = [i1: number, i2: number, n1: number, n2: number];

// 바이트 → 줄 목록. 줄마다 '\n' 을 품고, 마지막 줄만 없을 수 있다.
export function splitLines(data: Buffer): string[] {
  if (!data.length) return [];
  const parts = data.toString('latin1').split('\n');
  const last = parts.pop()!;
  const lines = parts.map((p) => p + '\n');
  if (last) lines.push(last);
  return lines;
}

// ── 2 단계: Myers 앞방향 탐욕 탐색 (SPEC.md §11.2) ────────────────

// 가운데 a·b 의 [ra, rb]. 대각선 k 마다 가장 멀리 간 x 를 V[k] 에.
//
// k 는 −d‥d 라 배열 첨자로 쓰려고 off 만큼 민다. d 마다 V 의 사본을
// 남겨 두었다가 (N, M) 에서 거꾸로 같은 판정을 되밟아 편집을 표시한다.
// O((N+M)·D) 시간, O((N+M)·D) 공간.
function forward(a: string[], b: string[]): [boolean[], boolean[]] {
  const [n, m] = [a.length, b.length];
  const off = n + m + 1;
  const v = new Int32Array(2 * off + 1);
  const trace: Int32Array[] = [];
  for (let d = 0; d <= n + m; d++) {
    trace.push(v.slice());
    for (let k = -d; k <= d; k += 2) {
      let x = k === -d || (k !== d && v[off + k - 1] < v[off + k + 1])
        ? v[off + k + 1]                   // 아래로: b 의 줄을 끼움
        : v[off + k - 1] + 1;              // 오른쪽: a 의 줄을 지움
      let y = x - k;
      while (x < n && y < m && a[x] === b[y]) [x, y] = [x + 1, y + 1];
      v[off + k] = x;
      if (x >= n && y >= m) return backtrack(trace, n, m, d, off);
    }
  }
  throw new Error('Myers 탐색이 끝나지 않았다');
}

function backtrack(trace: Int32Array[], n: number, m: number,
  dfin: number, off: number): [boolean[], boolean[]] {
  const ra = new Array<boolean>(n).fill(false);
  const rb = new Array<boolean>(m).fill(false);
  let [x, y] = [n, m];
  for (let d = dfin; d > 0; d--) {
    const v = trace[d];
    const k = x - y;
    const down = k === -d ||
      (k !== d && v[off + k - 1] < v[off + k + 1]);
    const pk = down ? k + 1 : k - 1;
    const px = v[off + pk];
    const py = px - pk;
    if (down) rb[py] = true;
    else ra[px] = true;
    [x, y] = [px, py];
  }
  return [ra, rb];
}

// 1·2 단계 — 앞뒤를 깎고 가운데를 앞방향 Myers 로.
export function myers(a: string[], b: string[]):
  [boolean[], boolean[]] {
  const [n, m] = [a.length, b.length];
  let s = 0;
  while (s < n && s < m && a[s] === b[s]) s++;
  let e = 0;
  while (e < n - s && e < m - s && a[n - 1 - e] === b[m - 1 - e]) e++;
  const [ra, rb] = forward(a.slice(s, n - e), b.slice(s, m - e));
  const pad = (k: number) => new Array<boolean>(k).fill(false);
  return [[...pad(s), ...ra, ...pad(e)], [...pad(s), ...rb, ...pad(e)]];
}

// ── 3 단계: 밀어 붙이기 (git 의 xdl_change_compact, 휴리스틱 없이) ─

// 바뀐 줄 묶음 [start, end). 빈 묶음(start == end)도 자리다. chg 끝에는
// 바뀌지 않은 가짜 줄이 하나 붙어 있다.
class Group {
  start = 0;
  end = 0;
  private n: number;

  constructor(private chg: boolean[]) {
    this.n = chg.length - 1;
    while (this.end < this.n && chg[this.end]) this.end++;
  }

  next(): boolean {
    if (this.end === this.n) return false;
    this.start = this.end = this.end + 1;
    while (this.end < this.n && this.chg[this.end]) this.end++;
    return true;
  }

  previous(): boolean {
    if (this.start === 0) return false;
    this.end = this.start = this.start - 1;
    while (this.start > 0 && this.chg[this.start - 1]) this.start--;
    return true;
  }

  slideDown(recs: string[]): boolean {
    if (this.end >= this.n || recs[this.start] !== recs[this.end]) {
      return false;
    }
    this.chg[this.start++] = false;
    this.chg[this.end++] = true;
    while (this.end < this.n && this.chg[this.end]) this.end++;
    return true;
  }

  slideUp(recs: string[]): boolean {
    const [s, e] = [this.start, this.end];
    if (s === 0 || recs[s - 1] !== recs[e - 1]) return false;
    this.chg[--this.start] = true;
    this.chg[--this.end] = false;
    while (this.start > 0 && this.chg[this.start - 1]) this.start--;
    return true;
  }
}

// 한 쪽 파일의 바뀐 줄 묶음을 밀어 자리를 정한다(SPEC.md §11.2).
//
// 묶음마다 위로 끝까지, 다시 아래로 끝까지 민다(밀다가 이웃 묶음과
// 붙으면 처음부터). 상대 파일의 바뀐 묶음과 끝이 맞는 자리가 있었으면
// 그리로 되올리고, 없으면 맨 아래에 둔다. 상대 쪽 묶음 표지 go 는
// 묶음과 발을 맞춰 움직인다 — 그래서 상대 쪽은 줄 내용이 아니라 바뀜
// 표시(ochg)만 있으면 된다. O(줄 수 × 미는 거리).
export function compact(recs: string[], rchg: boolean[],
  ochg: boolean[]): boolean[] {
  const chg = [...rchg, false];
  const g = new Group(chg);
  const go = new Group([...ochg, false]);
  for (;;) {
    if (g.end !== g.start) {
      let earliest: number;
      let matchEnd: number;
      let size: number;
      do {
        size = g.end - g.start;
        matchEnd = -1;
        while (g.slideUp(recs)) go.previous();
        earliest = g.end;
        if (go.end > go.start) matchEnd = g.end;
        while (g.slideDown(recs)) {
          go.next();
          if (go.end > go.start) matchEnd = g.end;
        }
      } while (size !== g.end - g.start);
      if (g.end !== earliest && matchEnd !== -1) {
        while (go.end === go.start) {
          g.slideUp(recs);
          go.previous();
        }
      }
    }
    if (!g.next()) break;
    go.next();
  }
  return chg.slice(0, -1);
}

// 세 단계를 다 거친 [ra, rb] — 계약의 전부(SPEC.md §11.2).
export function editFlags(a: string[], b: string[]):
  [boolean[], boolean[]] {
  let [ra, rb] = myers(a, b);
  ra = compact(a, ra, rb);
  rb = compact(b, rb, ra);
  return [ra, rb];
}

// 바뀐 곳들, 앞에서부터.
//
// 끝에서 앞으로 훑으며 같은 자리에서 만나는 지운 묶음과 끼운 묶음을
// 한 바뀐 곳으로 묶는다(git 의 xdl_build_script).
export function buildChanges(ra: boolean[], rb: boolean[]): Change[] {
  const out: Change[] = [];
  let [i1, i2] = [ra.length, rb.length];
  while (i1 > 0 || i2 > 0) {
    if ((i1 > 0 && ra[i1 - 1]) || (i2 > 0 && rb[i2 - 1])) {
      const [l1, l2] = [i1, i2];
      while (i1 > 0 && ra[i1 - 1]) i1--;
      while (i2 > 0 && rb[i2 - 1]) i2--;
      out.push([i1, i2, l1 - i1, l2 - i2]);
    } else {
      [i1, i2] = [i1 - 1, i2 - 1];
    }
  }
  return out.reverse();
}

// git 기본 드라이버의 함수 줄 — 첫 바이트가 영문자·'_'·'$'.
const isFunc = (line: string) => /^[A-Za-z_$]/.test(line);

// ASCII 공백만 뗀다(Python 의 bytes.rstrip). trimEnd 는 0xA0 도 떼는데,
// 바이트 문자열에서 0xA0 은 UTF-8 한 글자의 한 조각일 수 있다.
const rstrip = (s: string) => s.replace(/[ \t\n\v\f\r]+$/, '');

function span(start: number, count: number): string {
  const first = count ? start + 1 : start;
  return count === 1 ? `${first}` : `${first},${count}`;
}

// 덩어리들(SPEC.md §11.3). 같으면 ''.
export function unifiedDiff(a: string[], b: string[]): string {
  const ch = buildChanges(...editFlags(a, b));
  const out: string[] = [];
  for (let i = 0; i < ch.length;) {
    let j = i;
    while (j + 1 < ch.length &&
      ch[j + 1][0] - (ch[j][0] + ch[j][2]) <= 2 * CONTEXT) j++;
    const [first, last] = [ch[i], ch[j]];
    const s1 = Math.max(first[0] - CONTEXT, 0);
    const s2 = Math.max(first[1] - CONTEXT, 0);
    const e1 = Math.min(last[0] + last[2] + CONTEXT, a.length);
    const e2 = Math.min(last[1] + last[3] + CONTEXT, b.length);
    let func = '';
    for (let q = s1 - 1; q >= 0; q--) {
      if (isFunc(a[q])) {
        func = ' ' + rstrip(rstrip(a[q]).slice(0, 80));
        break;
      }
    }
    out.push(`@@ -${span(s1, e1 - s1)} +${span(s2, e2 - s2)} @@` +
      `${func}\n`);
    let p1 = s1;
    for (const [c1, c2, n1, n2] of ch.slice(i, j + 1)) {
      out.push(...a.slice(p1, c1).map((l) => ' ' + l),
        ...a.slice(c1, c1 + n1).map((l) => '-' + l),
        ...b.slice(c2, c2 + n2).map((l) => '+' + l));
      p1 = c1 + n1;
    }
    out.push(...a.slice(p1, e1).map((l) => ' ' + l));
    i = j + 1;
  }
  return out.map((line) => line.endsWith('\n') ? line
    : line + '\n\\ No newline at end of file\n').join('');
}

const q = (prefix: string, p: string) => quotePath(prefix + p);
const octal = (mode: number) => mode.toString(8).padStart(6, '0');

// 파일 하나의 diff 전체(SPEC.md §11.4). old·new 는 Side 또는
// null(새로 생김·지워짐). 경로는 바이트 문자열. 같으면 ''.
export function fileDiff(pathA: string, pathB: string, old: Side | null,
  nu: Side | null): string {
  if (old && nu && old[0] === nu[0] && old[1] === nu[1]) return '';
  const rows = [`diff --git ${q('a/', pathA)} ${q('b/', pathB)}`];
  const z = '0000000';
  if (old === null) {
    rows.push(`new file mode ${octal(nu![0])}`,
      `index ${z}..${nu![1].slice(0, 7)}`);
  } else if (nu === null) {
    rows.push(`deleted file mode ${octal(old[0])}`,
      `index ${old[1].slice(0, 7)}..${z}`);
  } else {
    if (old[0] !== nu[0]) {
      rows.push(`old mode ${octal(old[0])}`,
        `new mode ${octal(nu[0])}`);
    }
    if (old[1] === nu[1]) return rows.join('\n') + '\n'; // 모드만 바뀜
    let idx = `index ${old[1].slice(0, 7)}..${nu[1].slice(0, 7)}`;
    if (old[0] === nu[0]) idx += ` ${octal(old[0])}`;
    rows.push(idx);
  }
  const da = old ? old[2] : Buffer.alloc(0);
  const db = nu ? nu[2] : Buffer.alloc(0);
  const nameA = old ? q('a/', pathA) : '/dev/null';
  const nameB = nu ? q('b/', pathB) : '/dev/null';
  if (da.subarray(0, BINARY_PROBE).includes(0) ||
    db.subarray(0, BINARY_PROBE).includes(0)) {
    rows.push(`Binary files ${nameA} and ${nameB} differ`);
    return rows.join('\n') + '\n';
  }
  rows.push(`--- ${nameA}`, `+++ ${nameB}`);
  return rows.join('\n') + '\n' +
    unifiedDiff(splitLines(da), splitLines(db));
}

// [모드, 이름, 바이트] — 저장소의 blob 에서.
export function blobSide(gitdir: string, mode: number, oid: string):
  Side {
  return [mode, oid, objects.readObject(gitdir, oid)[1]];
}
